File: swift-bridge/Sources/OSAKitBridge/Component.swift
import Cocoa
import OSAKit

private func osaComponentInstancePointer(_ language: OSALanguage) -> UInt64 {
    UInt64(UInt(bitPattern: UnsafeMutableRawPointer(language.componentInstance)))
}

private func osaComponentSummary(_ language: OSALanguage) -> [String: Any] {
    [
        "componentInstancePointer": osaComponentInstancePointer(language),
        "language": osaLanguageInfo(language),
    ]
}

private func osaComponentInstanceSummary(_ handle: OSAComponentInstanceHandle) -> [String: Any] {
    [
        "componentInstancePointer": handle.componentInstancePointer,
        "language": osaLanguageInfo(handle.language),
    ]
}

@_cdecl("osa_component_from_language")
public func osa_component_from_language(_ languagePtr: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let languagePtr else { return nil }
    let language: OSALanguage = osaBorrow(languagePtr)
    return osaRetain(OSAComponentHandle(language))
}

@_cdecl("osa_component_from_language_instance")
public func osa_component_from_language_instance(_ instancePtr: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let instancePtr else { return nil }
    let instance: OSALanguageInstance = osaBorrow(instancePtr)
    return osaRetain(OSAComponentHandle(instance.language))
}

@_cdecl("osa_component_language")
public func osa_component_language(_ componentPtr: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let componentPtr else { return nil }
    let handle: OSAComponentHandle = osaBorrow(componentPtr)
    return osaRetain(handle.language)
}

@_cdecl("osa_component_summary_json")
public func osa_component_summary_json(
    _ componentPtr: UnsafeMutableRawPointer?,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutablePointer<CChar>? {
    guard let componentPtr else {
        osaWriteError(errorOut, "missing OSA component handle")
        return nil
    }
    let handle: OSAComponentHandle = osaBorrow(componentPtr)
    return osaCString(osaJSONString(osaComponentSummary(handle.language)))
}

@_cdecl("osa_component_instance_from_language")
public func osa_component_instance_from_language(_ languagePtr: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let languagePtr else { return nil }
    let language: OSALanguage = osaBorrow(languagePtr)
    return osaRetain(
        OSAComponentInstanceHandle(
            componentInstancePointer: osaComponentInstancePointer(language),
            language: language
        )
    )
}

@_cdecl("osa_component_instance_from_language_instance")
public func osa_component_instance_from_language_instance(_ instancePtr: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let instancePtr else { return nil }
    let instance: OSALanguageInstance = osaBorrow(instancePtr)
    return osaRetain(
        OSAComponentInstanceHandle(
            componentInstancePointer: UInt64(UInt(bitPattern: UnsafeMutableRawPointer(instance.componentInstance))),
            language: instance.language
        )
    )
}

@_cdecl("osa_component_instance_component")
public func osa_component_instance_component(_ instancePtr: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let instancePtr else { return nil }
    let handle: OSAComponentInstanceHandle = osaBorrow(instancePtr)
    return osaRetain(OSAComponentHandle(handle.language))
}

@_cdecl("osa_component_instance_summary_json")
public func osa_component_instance_summary_json(
    _ instancePtr: UnsafeMutableRawPointer?,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutablePointer<CChar>? {
    guard let instancePtr else {
        osaWriteError(errorOut, "missing OSA component instance handle")
        return nil
    }
    let handle: OSAComponentInstanceHandle = osaBorrow(instancePtr)
    return osaCString(osaJSONString(osaComponentInstanceSummary(handle)))
}

File: src/component.rs
use core::ffi::c_void;
use std::ptr;

use serde::{Deserialize, Serialize};

use crate::ffi;
use crate::language::{Language, LanguageSummary};
use crate::language_instance::LanguageInstance;
use crate::private::decode_json;
use crate::script_error::{from_swift, OsaKitError};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
/// Summarizes the Carbon OSA component backing an `OSALanguage`.
pub struct OsaComponentSummary {
    /// Holds the raw Carbon `ComponentInstance` pointer value reported by `OSAKit`.
    pub component_instance_pointer: u64,
    /// Captures the `OSALanguage` summary reported for the component.
    pub language: LanguageSummary,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
/// Summarizes the Carbon OSA component instance backing an `OSALanguageInstance`.
pub struct OsaComponentInstanceSummary {
    /// Holds the raw Carbon `ComponentInstance` pointer value reported by `OSAKit`.
    pub component_instance_pointer: u64,
    /// Captures the `OSALanguage` summary reported for the component.
    pub language: LanguageSummary,
}

#[derive(Debug)]
/// Wraps the Carbon OSA component resolved for an `OSALanguage`.
pub struct OsaComponent {
    raw: *mut c_void,
}

#[derive(Debug)]
/// Wraps the Carbon OSA component instance resolved for an `OSALanguageInstance`.
pub struct OsaComponentInstance {
    raw: *mut c_void,
}

impl OsaComponent {
    /// Resolves this wrapper from an `OSALanguage`.
    pub fn from_language(language: &Language) -> Result<Self, OsaKitError> {
        let raw = unsafe { ffi::osa_component_from_language(language.raw) };
        if raw.is_null() {
            return Err(OsaKitError::FrameworkError(
                "OSAKit could not resolve a Carbon component for the language".into(),
            ));
        }
        Ok(Self { raw })
    }

    /// Resolves this wrapper from an `OSALanguageInstance`.
    pub fn from_language_instance(instance: &LanguageInstance) -> Result<Self, OsaKitError> {
        let raw = unsafe { ffi::osa_component_from_language_instance(instance.raw) };
        if raw.is_null() {
            return Err(OsaKitError::FrameworkError(
                "OSAKit could not resolve a Carbon component for the language instance".into(),
            ));
        }
        Ok(Self { raw })
    }

    /// Returns the `OSALanguage` associated with this OSA component.
    pub fn language(&self) -> Result<Language, OsaKitError> {
        let raw = unsafe { ffi::osa_component_language(self.raw) };
        if raw.is_null() {
            return Err(OsaKitError::FrameworkError(
                "OSAKit returned a null language for the component".into(),
            ));
        }
        Ok(Language { raw })
    }

    /// Returns summary metadata for this OSA component.
    pub fn summary(&self) -> Result<OsaComponentSummary, OsaKitError> {
        let mut error_ptr = ptr::null_mut();
        let json = unsafe { ffi::osa_component_summary_json(self.raw, &mut error_ptr) };
        if json.is_null() {
            return Err(from_swift(ffi::status::FRAMEWORK_ERROR, error_ptr));
        }
        decode_json(json)
    }
}

impl Drop for OsaComponent {
    fn drop(&mut self) {
        if !self.raw.is_null() {
            unsafe { ffi::osa_object_release(self.raw) };
            self.raw = ptr::null_mut();
        }
    }
}

impl OsaComponentInstance {
    /// Resolves this wrapper from an `OSALanguage`.
    pub fn from_language(language: &Language) -> Result<Self, OsaKitError> {
        let raw = unsafe { ffi::osa_component_instance_from_language(language.raw) };
        if raw.is_null() {
            return Err(OsaKitError::FrameworkError(
                "OSAKit returned a null component instance for the language".into(),
            ));
        }
        Ok(Self { raw })
    }

    /// Resolves this wrapper from an `OSALanguageInstance`.
    pub fn from_language_instance(instance: &LanguageInstance) -> Result<Self, OsaKitError> {
        let raw = unsafe { ffi::osa_component_instance_from_language_instance(instance.raw) };
        if raw.is_null() {
            return Err(OsaKitError::FrameworkError(
                "OSAKit returned a null component instance for the language instance".into(),
            ));
        }
        Ok(Self { raw })
    }

    /// Returns the backing OSA component for this component instance.
    pub fn component(&self) -> Result<OsaComponent, OsaKitError> {
        let raw = unsafe { ffi::osa_component_instance_component(self.raw) };
        if raw.is_null() {
            return Err(OsaKitError::FrameworkError(
                "OSAKit returned a null component for the component instance".into(),
            ));
        }
        Ok(OsaComponent { raw })
    }

    /// Returns summary metadata for this OSA component instance.
    pub fn summary(&self) -> Result<OsaComponentInstanceSummary, OsaKitError> {
        let mut error_ptr = ptr::null_mut();
        let json = unsafe { ffi::osa_component_instance_summary_json(self.raw, &mut error_ptr) };
        if json.is_null() {
            return Err(from_swift(ffi::status::FRAMEWORK_ERROR, error_ptr));
        }
        decode_json(json)
    }
}

impl Drop for OsaComponentInstance {
    fn drop(&mut self) {
        if !self.raw.is_null() {
            unsafe { ffi::osa_object_release(self.raw) };
            self.raw = ptr::null_mut();
        }
    }
}
